### DISCUSSION TITLE
RFC: Removing JS/CSS Source Files From Releases

### DISCUSSION THREAD

**Discussion by ObliviousHarmony** [woocommerce/woocommerce#34212] [Request for Comments]
* Status: Rejected
* Closes: August 26th, 2022
* [Decision](https://github.com/woocommerce/woocommerce/discussions/34212#discussioncomment-3520367)

## Desired Feedback
Our goal in publishing this RFC is to gather feedback from our community about a change that will potentially impact developer debugging workflows. We are confident that our proposal is the right path forward, however, we would also like to make sure we’re taking every perspective into consideration. 

We will be accepting feedback for the next three weeks, after which we will make a decision about how to proceed. Please let us know if you see any issues with our solution or have suggestions about how we could better approach this problem.

## Problem
As a plugin’s file size increases, it becomes possible for the size of the ZIP archive to reach levels that make it difficult for users to update. On some hosting providers, timeouts can become a problem when it takes too long to download and unzip the plugin. **As we understand it, 10MB compressed seems to be the point at which this begins to become a problem**.

We’ve taken several smaller steps to decrease the size of our archive, however, it still continues to grow:

| Version          | ZIP Archive Size |
| ---------------- | ---------------- |
| WooCommerce 6.2  | 9.4 MB           |
| WooCommerce 6.3  | 9.4 MB           |
| WooCommerce 6.4  | 9.6 MB           |
| WooCommerce 6.5  | 9.6 MB           |
| WooCommerce 6.6  | 9.7 MB           |
| WooCommerce 6.7  | 9.9 MB           |

Given the work we’re doing in the backend, our React-based admin interface, and WooCommerce Blocks, we should expect this trend to continue. Based on the rate of growth we have seen this year, I expect we will be over 10 MB by WooCommerce 6.9 at the latest.

## Solution
According to [Plugin Guideline #4](https://developer.wordpress.org/plugins/wordpress-org/detailed-plugin-guidelines/#4-code-must-be-mostly-human-readable), we don’t need to include any source files as long as we are linking to our GitHub repository. This is further clarified by [an article published by the Plugin Review Team](https://make.wordpress.org/plugins/2018/08/23/guideline-update-clarifications-to-trialware-and-human-readability/). **I propose that we remove all JavaScript and CSS source files from the plugin we distribute**:

* All JS, such as our React-based admin interface and jQuery interfaces, will be included in their transpiled form, both minified and mangled.
* All CSS will be included, but, preprocessor source files like `*.scss` will be excluded.

In addition to this, we have discussed the possibility of releasing a plugin containing source files and maps for all of the JS and CSS files. This could be installed alongside WooCommerce and would enable debugging in WordPress. This may not be necessary, but, we wanted to highlight it as an option for including source files without having them in the plugin we ship to all merchants.

## Benefit
Decreasing the size of our archive will decrease the likelihood of there being size-related update problems. I (naively) deleted source files from the WooCommerce 6.7 plugin archive and managed to get a 30% decrease in compressed size. With a more deliberate approach to optimizing post-build file sizes, I’m sure there’s a lot we can do to decrease this further.

## Consequence
This change will primarily affect the accessibility of local debugging by end-users. We currently provide support for the `SCRIPT_DEBUG` constant. When active, this serves source(-ish) versions of our JavaScript and CSS files. It’s worth noting that _these are still build tool outputs_, meaning they aren’t really source files. Honestly, it’s not that much different than the browser’s automatic formatting when viewing minified files.

The other concern is transparency. My proposal includes mangling first-party JavaScript, which makes it entirely unreadable to humans. You could clone our repository and build these files to validate the output, but, it’s still an obstacle.

### Source-Included Build
We could optionally provide a build that includes the source files. This would allow us to provide a streamlined build as well as one containing all of the source JS and CSS files. To be honest, this sounds like developers should just use the repository. It would be great to hear some of the value that supporting this use case might provide.

### Minification Only
In response to potential transparency concerns, one alternative could be to instead only minify the JavaScript. This would come with an increase in the archive size, however, it would mean that the JavaScript is more human-readable on the filesystem.

**Comment by siliconforks** [comment]
> This change will primarily affect the accessibility of local debugging by end-users. We currently provide support for the `SCRIPT_DEBUG` constant. When active, this serves source(-ish) versions of our JavaScript and CSS files.

I don't think this is entirely accurate?  I believe the unminified CSS files were removed from the distributed plugin `*.zip` file a long time ago - the `SCRIPT_DEBUG` constant currently only affects `*.js` files (and only a subset of the `*.js` files that come with the distributed plugin `*.zip` file).

Why not just keep all the unminified JavaScript files that are currently used by the `SCRIPT_DEBUG` constant and exclude everything else that is not needed from the plugin `*.zip` file?  (That is, exclude all `*.scss` files, all `*.ts` files, and any `*.js` files that aren't used directly or with `SCRIPT_DEBUG`).  This would probably give 90% of the size reduction while keeping `SCRIPT_DEBUG` working as it currently does.

**Reply by ObliviousHarmony** [reply]
> I believe the unminified CSS files were removed from the distributed plugin *.zip file a long time ago

You're right, this does seem to be the case. All of the `*.css` files are minified. I apologize for the confusion.

> Why not just keep all the unminified JavaScript files that are currently used by the SCRIPT_DEBUG constant and exclude everything else that is not needed from the plugin *.zip file?

For quite some time now, [WooCommerce Admin](https://wordpress.org/plugins/woocommerce-admin/) and [WooCommerce Blocks ](https://woocommerce.com/products/woocommerce-gutenberg-products-block/) have been included within WooCommerce Core. While support for `SCRIPT_DEBUG` here is inconsistent, it is still present in some cases. If we're going to say that `SCRIPT_DEBUG` support is important, how do you envision us supporting that as we move towards creating a more modern frontend for both customers and merchants?

I feel like this is one of the few cases where a binary policy makes the most sense. Could you expand a bit on what value you're getting out of `SCRIPT_DEBUG` as opposed to a different approach to debugging?

**Reply by siliconforks** [reply]
In my mind, the two requirements are these:

1. There needs to be an easy way to switch a WordPress/WooCommerce site into a "debugging mode" where it is possible to debug using unminified JavaScript.
2. Ideally, requirement 1 should be implemented with some sort of "standard" or conventional method, preferably one which works for WordPress core, themes, and plugins.  Right now this is the `SCRIPT_DEBUG` constant.

The really critical requirement is requirement 1.  Requirement 2 is less important - if there were some other method available like using a special debugging plugin, then that might be a viable alternative.

**Comment by daigo75** [comment]
Without readable source files, debugging an issue could become impossible. That's why I always include source maps, especially for TypeScript. One doesn't necessarily have access to the site to upload a "developer friendly" version of the code (actually, I don't remember the last time I **did** have access). On top of that, I wouldn't go to the extent of cloning a site and install additional debugging tools to troubleshoot something that could be a minor JS conflict. Much easier to be able to inspect it on the live site itself. 

Having an open source project, but without the source when one needs it, could make things very tricky to handle.

**Reply by helgatheviking** [reply]
>Without readable source files, debugging an issue could become impossible.

THIS! It's already getting impossible for me to work on the repo version, but now I can't even debug my clients' sites?

**Reply by daigo75** [reply]
One of the issues I encountered with the repo version is that it takes a specific setup to build, and the built version may or may not be identical to the one running on a client's site. The discrepancies could be minor, but still enough to cause a mismatch between the code that one can inspect (repo) and the one running on the client's site, which is the one that needs to be investigated.

**Comment by pexlechris** [comment]
What if you downloaded files in patches in a cronjob, after Woocommerce installs, such as Updating Database Tables procedure?

**Comment by NTShop** [comment]
I agree with @helgatheviking, building a repo version is difficult and time consuming to say the least. I also agree with @daigo75, very often I find myself working on versions that are not up to date and people have various reason for postponing an update. That is an important point to keep in mind. 

I have a plugin that is used on many hundreds of sites, it's 12MB in size and has its own update mechanism in it, only once in 10 years has anyone every told me they couldn't update because of the file size or timeout, and that was because of overly restrictive settings in their PHP config, which they remedied quickly.

Given that WooCommerce is so widely used, I would expect that most hosting providers would accommodate it's zip file size instead of WooCommerce core developers having to downsize or limit the zip to suit hosting providers. That just seems backwards to me. Plus it's not like it's growing rapidly by a large amount.

**Comment by scottbuscemi** [comment]
"On some hosting providers, timeouts can become a problem when it takes too long to download and unzip the plugin. As we understand it, 10MB compressed seems to be the point at which this begins to become a problem."

Are there stats or more information about this available?

**Reply by ObliviousHarmony** [reply]
Sorry @scottbuscemi, this would be a question for the WordPress team.

**Comment by seb86** [comment]
Why not just leave the source files on the repo? You can remove the source files that don't require debugging for the public release and allow developers to install a developer version of WooCommerce via Composer. That's basically what I have done for my plugin even though it's not even 1MB yet, the less space used on a live site from the many plugins we install the better.

**Reply by scottbuscemi** [reply]
That's the idea.

**Reply by helgatheviking** [reply]
>install a developer version of WooCommerce via Composer

Wondering how you manage this... do you get SSH access for every customer who uses your plugin?

**Comment by susameca** [comment]
Without readable source files, debugging an issue could become impossible. I also use them for debugging and developing new features.

**Comment by helgatheviking** [comment]
>we have discussed the possibility of releasing a plugin containing source files and maps for all of the JS and CSS files. This could be installed alongside WooCommerce and would enable debugging in WordPress.

This might be a viable compromise. I think we _need_ a way to have source files running on a live/staging site. *finding* the issue is usually more difficult than fixing it! Won't this make debugging difficult for your own support people?

But installing a plugin is doable with only admin user access and wouldn't require I get SSH for all my plugin customers who need debugging help. 

What about a script debug setting in WC core? Enabling it could use the action scheduler to download the source files on a delay....

**Reply by ObliviousHarmony** [reply]
Thanks for pointing this out @helgatheviking, I think it's a key point in the debugging-oriented concerns. I likely could have done a better job of highlighting this possibility.

> What about a script debug setting in WC core? Enabling it could use the action scheduler to download the source files on a delay

I was thinking it could be in the "WooCommerce > Status > Tools" menu or something and it would let you install it from there. I see some value in an all-encompassing WooCommerce Debugging extension that supplies source files and potentially other tools if there's some need for it.

**Reply by helgatheviking** [reply]
I def missed it the first read-through.  :) If it's absolutely necessary, I think that's workable. It will introduce an additional step to debugging (ex: viewing a customer's page in the console may not reveal anything until the additional plugin is installed) but it's not insurmountable.

**Comment by Lumieredelune** [comment]
Why not split your plugin into two or three, required by the first one ? It would make smaller .zip files

`WooCommerce Admin and WooCommerce Blocks`

Maybe you could make them separate again ? That's the road followed by many plugins and themes and it makes life easier than an impossible-to-debug-plugin.

**Reply by ObliviousHarmony** [reply]
One of the significant problems with this approach is that these are not extensions to the platform but rather iterations on the platform itself. WooCommerce is _very_ coupled internally and does not expose the kind of code API needed to implement most of these features. Many extensions are filled with hacks and workarounds to get at WooCommerce internals that should just be accessible in the first place.

By developing these features directly within WooCommerce we can make platform-level improvements without having to work around the many problems introduced by version compatibility mismatches. This does not mean introducing breaking changes, but, it lets us iteratively develop better solutions that improve the performance and usability of WooCommerce as a whole.

**Comment by Marc-pi** [comment]
Hello,
i am not able to upgrade to 6.8 from 6.7 (from WP Admin), it is the first time the upgrade process fails ;-)
it crashed woo in the installed extension list, i fixed by reinstalling after dir deletion, but i fear only few people know how to fix...
is there a special (alternative) procedure to upgrade ?

**Comment by digamber89** [comment]
If we are going to go - to the plugin as a debugger route - i would ask that there be a message in the console when SCRIPT_DEBUG is set to true. Something like - if you're looking to debug WooCommerce - please add the debugger plugin - that way new developers won't spend hours trying to make sense of jumbled up javascript

**Comment by Marc-pi** [comment]
> i would ask that there be a message in the console when SCRIPT_DEBUG is set to true

yep, I proposed a similar solution there https://github.com/woocommerce/woocommerce/issues/34281
but some areas are in the making at WP Core (see the updating-the-updaters feature project url link I posted in my bug declaration)
Actually, WC disapears completely from installed plugins when upgrade aborts. It is surprising for non experts !

**Comment by d9media** [comment]
Remove them and allow to side-load them from within the WooCommerce settings, maybe status tab or introduce a developer tab that can be activated from the status tab or by env_var

**Comment by BusinessBloomer** [comment]
My 2 cents - WooCommerce blocks and WooCommerce admin should not be in core, as well as hundreds of lines of code that advertise Jetpack and integrate with other extensions.

If we are to build a lean WooCommerce and make it more performant and easy to use, these should be kept out of the main "skeleton". 

Most websites don't require JS powered analytics (and those who do, use third party tracking & reporting because Woo admin is very difficult or impossible to customize) and many more don't need 20+ blocks (we could just "install" a single block?), as well as ads or integrations.

So, if size is a problem, I'd simply stop adding to it and optimize what's there, as well as removing the items I mentioned

**Reply by helgatheviking** [reply]
Not to mention, the Woo admin is wicked slow (at least locally when doing dev work). 🐢

**Reply by ObliviousHarmony** [reply]
I think [this reply](https://github.com/woocommerce/woocommerce/discussions/34212#discussioncomment-3352150) adequately covers the rationale for including both of these within WooCommerce. It's worth mentioning too; I only used the WooCommerce Admin name to more clearly reference the code I was talking about. The WooCommerce Admin feature plugin no longer exists and has been entirely folded into WooCommerce.

**Comment by dossy** [comment]
I propose that the mainline plugin should be strictly a "release build" that only contains the minimal amount of assets absolutely necessary and required to run in production, which should be as light-weight as possible, minified, mangled, etc.

A separate debug plugin should be created which will side-load any and all (and only) debug assets.  This would be completely optional, installable for troubleshooting and for development instances as necessary.

The build process in the GitHub repo should create the packages for both of these plugins from the original project source files.

Using WordPress hooks and filters, side-loading the debug plugin and then conditionally enabling parts of it as desired from settings should be clean.

**Comment by shahidbahader** [comment]
I think the plugin should be divided into 3 plugin, 
Woocommerce
Woocommerce Admin
Woocommerce Block

This will reduce the size of the plugin

**Comment by ObliviousHarmony** [comment]
First of all, I would like to thank everyone for their input on this RFC. It has been really nice seeing how many have participated in the discussion and we look forward to being able to do more of this in the future.

Based on the feedback that we've received here, we have decided against moving forward with this proposal. While it is something we may need to address in the future, until it proves to be a problem, we will leave things as they are.
